### PR TITLE
Issue #5953: Add the initial Jenkinsfile for the project

### PR DESCRIPTION
**Pull request by daniilyar**
Issue #5953: add the initial Jenkinsfile for the project

### CODE REVIEW

**Comment by romani** [comment]
Please move content of Jenkins.md to wiki.
No need of such content in repo

**Comment by romani** [comment]
Please fix commit message

Please update all other CI configs to skipped Jenkins file

**Comment by daniilyar** [comment]
@romani , I have added Jenkinsfile to the ignorelist of 4 CIs:
- wercker
- shippable
- appveyor
- travis
Could you please confirm that I didn't miss anyone else? I checked other CI files in the repo, and didn't found any skip lists except aforesaid ones. Thanks!

**Comment by daniilyar** [comment]
@romani, the build now fails because of:
```
New misspellings found, please review:
Jenkinsfile
xterm
```
Both are known words that can't be changed because both are used as a keywords in the script that will stop working if these words will be changed.

The error does not state how do I add them to ignore list, is there an option to do so?

**Comment by romani** [comment]
> Both these are known words that can't be changed because both are used as a keywords in the script that will stop working if these words will be changed.

just add this words to whiltelist as suggested at https://travis-ci.org/checkstyle/checkstyle/jobs/395923279#L672

> Probably, it doesn't work as expected. Could you please help me to understand what I am doing wrong?

yes, you added it to skip, but :) you changes not only Jenkins file, you changes travis file so skip does not work. Additionally "spell checker" is running all the time - https://github.com/checkstyle/checkstyle/blob/master/.travis.yml#L208 to keep spelling under control for all changes :). I know it is hard to write a good code :), it is struggle even for me.

File: .travis.yml [hunk 1]
```diff
@@ -211,7 +211,8 @@ script:
   - SKIP_FILES1=".github|codeship-*|buddy.yml|appveyor.yml|circleci|distelli-manifest.yml"
   - SKIP_FILES2="|fast-forward-merge.sh|LICENSE|LICENSE.apache20|README.md|release.sh|RIGHTS.antlr"
   - SKIP_FILES3="|shippable.yml|shippable.sh|wercker.yml|wercker.sh|intellij-idea-inspections.xml"
-  - SKIP_FILES=$SKIP_FILES1$SKIP_FILES2$SKIP_FILES3
+  - SKIP_FILES4="|Jenkinsfile"
+  - SKIP_FILES=$SKIP_FILES1$SKIP_FILES2$SKIP_FILES3$SKIP_FILES4
```
Comment (romani) [inline .travis.yml]: please add it to all CIs, you missed:
https://github.com/checkstyle/checkstyle/blob/master/.circleci/config.yml#L21  (pitest)
https://github.com/checkstyle/checkstyle/blob/master/distelli-manifest.yml#L9  (javadoc regression)

Just FYI we also have : buddy.yml, codeship-steps.yml , but such CIs are limited for `master` only , so I did not have much time to update them too to skip builds, so for now ok to skip them in update.

**Comment by daniilyar** [comment]
> please make it Issue #5953: add the initial Jenkinsfile , commit should be read as "[It will] add the initial Jenkinsfile" , so imperative form is required.

Not sure I understand what to change. Commit already states `Issue #xxx: Add the initial Jenkinsfile`, exactly as you requested. The only difference is that I included the information that it is a part one (i.e. the initial setup only). But it is still writeen in the imperative form. I tend to follow the GitHub recommendations on commit messages, so all my commits are always written in such a form - '[will] Add' instead of 'Added' and so on..

> Commit message will not be used in release generation is issue is referenced
sorry, could you explain what did you mean here?

File: .travis.yml (hunk 1 above)
Comment (daniilyar) [reply .travis.yml]: Done, thanks!

**Comment by rnveach** [comment]
> > Commit message will not be used in release generation is issue is referenced

> sorry, could you explain what did you mean here?

Release notes is based on issue title, not commit message or PR title. We only use commit message to identify the issue being referenced.

> Not sure I understand what to change.

The only difference I am seeing is to make `Add` to `add` (lowercase `a`). You removed the `Part 1` which is I believe is what he wanted removed.

**Comment by rnveach** [comment]
@daniilyar Travis is complaining about spellchecker. I think the words are out of order in the whitelist file.
````
@@ -622,0 +623 @@
+Jenkinsfile
@@ -1311,0 +1313 @@
+xterm
@@ -1332,2 +1333,0 @@
-Jenkinsfile
-xterm
````

**Comment by daniilyar** [comment]
> Release notes is based on issue title, not commit message or PR title. We only use commit message to identify the issue being referenced.
Got it, thanks!

> the words are out of order in the whitelist file.
So, it should be alphabetically sorted? Got it, thanks! 
Probably I missed the documentation for the spellchecker, sorry.. Will fix today :)

**Comment by rnveach** [comment]
You can always run it locally to confirm it will pass. Just run `./.ci/test-spelling-unknown-words.sh`.

**Comment by daniilyar** [comment]
> You can always run it locally to confirm it will pass. Just run ./.ci/test-spelling-unknown-words.sh.
@rnveach, I already tried to run it locally, but the script failed with:
```
./test-spelling-unknown-words.sh: line 18: .ci/jsoref-spellchecker/words.rpm: No such file or directory
```
Not sure why it requires rpm, on my vanilla Debian linux rpm packages are not supported, maybe this is the reason for failure, I didn't have time to investigate it.

**Comment by daniilyar** [comment]
@romani, I have addressed all your suggestions. Build now passes well, please check it again when you have a minute. Thanks

File: .circleci/config.yml [hunk 2]
```diff
@@ -19,6 +19,7 @@ jobs:
             echo -n "|shippable.sh|wercker.yml|wercker.sh" >> .ci-temp/skip_ci_files
             echo -n "|intellij-idea-inspections.xml" >> .ci-temp/skip_ci_files
             echo -n "|org.eclipse.jdt.core.prefs" >> .ci-temp/skip_ci_files
+            echo -n "Jenkinsfile" >> .ci-temp/skip_ci_files
```
Comment (romani) [inline .circleci/config.yml]: | is missed

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  agent { node { label "${JENKINS_NODE}" } }
+
+  // Various build and build plugin's options
+  options {
+    ansiColor('xterm')
+  }
+
+  stages {
+
+    stage("Build") {
```
Comment (romani) [inline Jenkinsfile]: as next PR, please update jenkins file to have ability to run several commands and some of them in parallel.

Comment (daniilyar) [reply Jenkinsfile]: @romani sure. I have already prepared the code change that runs multipke commands in Hyper.sh in parallel, but to test it I need this PR to be merged :)

File: .circleci/config.yml (hunk 2 above)
Comment (daniilyar) [reply .circleci/config.yml]: Fixed, thanks a lot!

**Comment by romani** [comment]
> system ID http://maven.apache.org/xsd/decoration-1.0.0.xsd:

we had such problem few times in past each time it self resolved, might be temporal problem with connection to such resource. Is it become a problem we can copy such xsd to some our location and configure xml validation to use our copy (in catalog file).

**Comment by daniilyar** [comment]
@romani , after re-start, all CIs passed well

**Pull request merged**